jumanji/environments/swarms/search_and_rescue/env.py

```diff
@@ -0,0 +1,390 @@
+            generator: Initial state `Generator` instance. Defaults to `RandomGenerator`
+                with 20 targets and 10 searchers.
+        """
+        self.searcher_vision_range = searcher_vision_range
+        self.target_contact_range = target_contact_range
+        self.num_vision = num_vision
+        self.agent_radius = agent_radius
+        self.searcher_params = AgentParams(
+            max_rotate=searcher_max_rotate,
+            max_accelerate=searcher_max_accelerate,
+            min_speed=searcher_min_speed,
+            max_speed=searcher_max_speed,
+            view_angle=searcher_view_angle,
+        )
+        self.max_steps = max_steps
+        self._viewer = viewer or SearchAndRescueViewer()
+        self._target_dynamics = target_dynamics or RandomWalk(0.01)
+        self.generator = generator or RandomGenerator(num_targets=20, num_searchers=10)
+        super().__init__()
+
+    def __repr__(self) -> str:
+        return "\n".join(
+            [
+                "Search & rescue multi-agent environment:",
+                f" - num searchers: {self.generator.num_searchers}",
+                f" - num targets: {self.generator.num_targets}",
+                f" - search vision range: {self.searcher_vision_range}",
+                f" - target contact range: {self.target_contact_range}",
+                f" - num vision: {self.num_vision}",
+                f" - agent radius: {self.agent_radius}",
+                f" - max steps: {self.max_steps},"
+                f" - target dynamics: {self._target_dynamics.__class__.__name__}",
+                f" - generator: {self.generator.__class__.__name__}",
+            ]
+        )
+
+    def reset(self, key: chex.PRNGKey) -> Tuple[State, TimeStep[Observation]]:
+        """Initialise searcher positions and velocities, and target positions.
+
+        Args:
+            key: Random key used to reset the environment.
+
+        Returns:
+            state: Initial environment state.
+            timestep: TimeStep with individual search agent views.
+        """
+        state = self.generator(key, self.searcher_params)
+        timestep = restart(observation=self._state_to_observation(state))
+        return state, timestep
+
+    def step(self, state: State, actions: chex.Array) -> Tuple[State, TimeStep[Observation]]:
+        """Environment update.
+
+        Update searcher velocities and consequently their positions,
+        mark found targets, and generate rewards and local observations.
+
+        Args:
+            state: Environment state.
+            actions: Arrays of searcher steering actions.
+
+        Returns:
+            state: Updated searcher and target positions and velocities.
+            timestep: Transition timestep with individual agent local observations.
+        """
+        searchers = update_state(state.key, self.searcher_params, state.searchers, actions)
+        key, target_key = jax.random.split(state.key, num=2)
```

Comment:
Looks like you're re-using `key` quite a few times, be sure to split up here for as many times as you need random calls. Also I'd recommend doing this at the top of the function and then pass one of the split keys into `update_state`

Reply:
Apart from the key passed to the target, all the functionality is deterministic, so all these keys are just dummy values that don't really need splitting. But yes you are correct, I should split at the top. I'll add a note to make this clear.